README


This archive contains the distribution Bloomd-Client,
version 0.25:

  Perl client to the bloomd server

This software is copyright (c) 2013 by Damien "dams" Krotkine.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
{{$NEXT}}

0.27      2014-10-03 14:16:15 Europe/Amsterdam

0.26      2014-03-05 14:00:05 Europe/Amsterdam

0.25      2013-10-03 14:34:08 Europe/Paris

0.24      2013-10-01 17:30:36 Europe/Paris
  * fix issue #11, parameters are now validated (Tiago Peczenyj)

0.23      2013-08-09 20:46:17 Europe/Paris

  * fix issue #9: perl 5.12 not required, require 5.10

0.22      2013-06-18 22:58:23 Europe/Paris

0.21      2013-06-12 22:22:57 Europe/Paris

0.20      2013-06-07 11:36:36 Europe/Paris

0.19      2013-06-07 11:24:37 Europe/Paris
  * fix compile tests

0.18      2013-06-07 10:59:22 Europe/Paris

0.17      2013-06-07 10:57:47 Europe/Paris
  * fix minimum perl version requirement for cpan testers

0.16      2013-06-07 00:05:59 Europe/Paris
  * Fix Travis and dist.ini

0.15      2013-06-06 23:58:05 Europe/Paris
  * Minor fixes
  * Configure Travis for this project

0.14      2013-06-06 18:27:37 Europe/Paris

0.13      2013-06-06 14:09:10 Europe/Paris

0.12      2013-06-05 17:19:02 Europe/Paris

0.11      2013-06-05 15:15:29 Europe/Paris

0.10      2013-06-05 15:00:35 Europe/Paris
  * First implementation
  

t/tlib/TestTimeout.pm
package TestTimeout;

use feature ':5.12';

use FindBin qw($Bin);
use lib "$Bin/../../lib";

use Test::More;
use Test::Exception;
use Test::MockObject;
use Bloomd::Client;
use Test::TCP;
use POSIX qw(ETIMEDOUT ECONNRESET strerror);
use Exporter 'import';

require bytes;

@EXPORT_OK = qw(test_timeout test_normal_wait);

sub create_server_with_timeout {
    my $in_timeout = shift;

    Test::TCP->new(
        code => sub {
            my $port   = shift;
            my $socket = IO::Socket::INET->new(
                Listen    => 5,
                Timeout   => 1,
                Reuse     => 1,
                Blocking  => 1,
                LocalPort => $port
            ) or die "ops $!";

            my $message = pack( 'c', 2 );

            my $buffer;
            while (1) {
                my $client = $socket->accept();

                sleep($in_timeout) if $in_timeout;
                if ( my $line = $client->getline( ) ) {
                    say STDERR " --- DEBUG line [$line]";
                    $client->print('Done');
                    sleep(1);
                }

                $client->close();
            }
        },
    );
}

sub test_timeout {
    my $provider = shift;

    subtest
      "Test timeout provider $provider, when the server wait 2 seconds to send the response"
      => sub {
        plan tests => 2;

        my $server = create_server_with_timeout(2);

        my $client = Bloomd::Client->new(
            host             => '127.0.0.1',
            port             => $server->port,
            timeout          => 0.1,
        );

        my $etimeout = strerror(ETIMEDOUT);
        my $ereset   = strerror(ECONNRESET);
        throws_ok { $client->list() } qr/Error in 'ping' : $etimeout/,
          "should die in case of timeout";
        throws_ok { $client->ping() }
        qr/Error in 'ping' : $ereset/,
          "using provider $provider_name, should close the connection";
      };
}

# sub test_normal_wait {
#     my $provider = shift;
#     my $timeout  = shift;

#     $timeout //= 0;
#     my $provider_name = $provider // "undef";

#     subtest
#       "Test timeout provider $provider_name when the server wait $timeout (timeout 2 seconds, should not die)"
#       => sub {
#         plan tests => 1;

#         my $server = create_server_with_timeout($timeout);

#         my $client = Riak::Light->new(
#             host             => '127.0.0.1',
#             port             => $server->port,
#             timeout          => 2,
#             timeout_provider => $provider
#         );

#         $provider //= "none";

#         lives_ok { $client->ping() } "using provider $provider, should wait";
#       };
# }

1;


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601",
    "Test::Exception" => 0,
    "Test::More" => 0,
    "Test::TCP" => 0
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Perl client to the bloomd server",
  "dist_author" => [
    "Damien \"dams\" Krotkine"
  ],
  "dist_name" => "Bloomd-Client",
  "dist_version" => "0.25",
  "license" => "perl",
  "module_name" => "Bloomd::Client",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Config" => 0,
    "IO::Socket::INET" => 0,
    "List::MoreUtils" => 0,
    "Method::Signatures" => 0,
    "Moo" => 0,
    "Type::Tiny" => "0.026",
    "perl" => "5.010"
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Bloomd/Client.pm
package Bloomd::Client;

# ABSTRACT: Perl client to the bloomd server

use feature ':5.10';
use Moo;
use List::MoreUtils qw(any mesh);
use Carp;
use Socket qw(:crlf);
use IO::Socket::INET;
use Errno qw(:POSIX);
use POSIX qw(strerror);
use Config;
use Types::Standard -types;

=attr protocol

The protocol. ro, defaults to 'tcp'

=cut

has protocol => ( is => 'ro', default => sub {'tcp'} );

=attr host

The host to connect to. ro, defaults to '127.0.0.1'

=cut

has host => ( is => 'ro', isa => StrMatch[qr/.+/], default => sub {'127.0.0.1'} );

=attr port

The port to connect to. ro, defaults to '8673'

=cut

has port => ( is => 'ro', isa => Int, default => sub {8673} );
has _socket => ( is => 'lazy', predicate => 1, clearer => 1 );

=attr timeout

The timeout (on read and write), in seconds. Can be a float. ro, defaults to 10.

=cut

has timeout => ( is => 'ro', , default => sub { 10 } );

has '_pid' => (is => 'ro', lazy => 1, clearer => 1, default => sub { $$ });

=head1 DESCRIPTION

This is a Perl client to connect to the Bloomd server. See
L<http://armon.github.io/bloomd/> for the server website.

=head1 SYNOPSIS

  use feature ':5.12';
  use Bloomd::Client;
  my $b = Bloomd::Client->new( timeout => 0.2 );
  my $filter = 'test_filter';
  $b->create($filter);
  my $array_ref = $b->list();
  my $hash_ref = $b->info($filter);
  $b->set($filter, 'u1');
  if ($b->check($filter, 'u1')) { say "it exists!" }
  my $hashref = $b->multi( $filter, qw(u1 u2 u3) );

=cut

sub _build__socket {
    my ($self) = @_;
    my $socket = IO::Socket::INET->new(
        Proto => $self->protocol,
        PeerHost => $self->host,
        PeerPort => $self->port,
        Timeout  => $self->timeout,
    ) or die "Can't connect to server: $!";

    $self->timeout
      or return $socket;

    $Config{osname} eq 'netbsd' || $Config{osname} eq 'solaris'
      and croak "the timeout option is not yet supported on NetBSD or Solaris";

    my $seconds  = int( $self->timeout );
    my $useconds = int( 1_000_000 * ( $self->timeout - $seconds ) );
    my $timeout  = pack( 'l!l!', $seconds, $useconds );

    $socket->setsockopt( SOL_SOCKET, SO_RCVTIMEO, $timeout )
      or croak "setsockopt(SO_RCVTIMEO): $!";

    $socket->setsockopt( SOL_SOCKET, SO_SNDTIMEO, $timeout )
      or croak "setsockopt(SO_SNDTIMEO): $!";

    return $socket;

}

=method disconnect

Closes the connection and reset the internal socket

=cut

sub disconnect {
    my ($self) = @_;
    $self->_has_socket
      and $self->_socket->close;
    $self->_clear_socket;
}

=method create

Creates a new bloom filter

  $b->create($name, $capacity?, $probability?, $in_memory=0|1?)

Only the name is mandatory. You can specify the initial capacity (if not the
filter will be enlarged as needed), the probability of maximum false positives
you want, and a flag to force the filter to not be in memory (by default it is).

Returns true if the filter didn't exist and was created successfully.

=cut

sub create {
    my ($self, $name, $capacity, $prob, $in_memory) = @_;
    my $args =
        ( $capacity ? "capacity=$capacity" : '' )
      . ( $prob ? " prob=$prob" : '' )
      . ( $in_memory ? " in_memory=$in_memory" : '' );
    $self->_execute("create $name $args" ) eq 'Done';
}

=method list

  my $array_ref = $b->list($prefix?)

Returns an ArrayRef of the existing filters. You can provide a prefix to filter
on them.

=cut

sub list {
    my ($self, $prefix) = @_;
    $prefix //= '';
    my @keys = qw(name prob size capacity items);
    [
     map {
         my @values = split / /;
         +{ mesh @keys, @values };
     }
     $self->_execute("list $prefix" )
    ];
}

=method drop

  $b->drop($name)

Drops a filter. Returns true if the filter existed and was removed properly.

=cut

sub drop {
    my ($self, $name) = @_;
    $self->_execute("drop $name") eq 'Done';
}


=method close

  $b->close($name)

Closes a filter. Returns true on success.

=cut

sub close {
    my ($self, $name) = @_;
    $self->_execute("close $name") eq 'Done';
}

=method clear

  $b->clear($name)

Clears a filter. Returns tru on success.

=cut

sub clear {
    my ($self, $name) = @_;
    $self->_execute("clear $name") eq 'Done';
}

=method check

  if ($b->check($name, $key)) {
    print "the element $key matched\n";
  }

Given a filter name and a key name, returns true if the key was previously
added in the filter (using C<set>).

=cut

sub check {
    my ($self, $name, $key) = @_;
    $self->_execute("c $name $key") eq 'Yes';
}

=method multi

  my $hash_ref = $b->multi($name, $key1, key2, key3);

Given a filter name and a list of elements, returns a HashRef, which keys are
the elements, and the values are 1 or 0, depending if the element is present in
the filter or not.

=cut

sub multi {
    my ($self, $name, @keys) = @_;
    @keys
      or return {};
    my @values = map { $_ eq 'Yes' } split / /, $self->_execute("m $name @keys");
    +{mesh @keys, @values };
}

=method set

  $b->set($name, $key);

Adds the element to the given filter. Returns 1 if the elements was not
previously in the filter and was properly added.

=cut

sub set {
    my ($self, $name, $key) = @_;
    $self->_execute("s $name $key") eq 'Yes';
}

=method bulk

  $b->bulk($name, $key1, $key2, $key3);

Adds the elements to the given filter. Returns void

=cut

sub bulk {
    my ($self, $name, @keys) = @_;
    @keys
      or return;
    $self->_execute("b $name @keys");
    return;
}

=method info

  my $hash_ref = $b->info($name);

Returns a HashRef giving informations about the given filter name.

=cut

sub info {
    my ($self, $name) = @_;
    +{ map { split / / } $self->_execute("info $name") };
}

=method flush

  $b->flush($name);

Flushes the filter. Returns true on success.

=cut

sub flush {
    my ($self, $name) = @_;
    $self->_execute("info $name") eq "Done";
}

sub _execute {
    my ($self, $command) = @_;
	if ($self->_pid ne $$) {
		$self->_clear_socket;
		$self->_clear_pid;
	}
    my $socket = $self->_socket;

    local $\;
    $socket->print($command . $CRLF)
      or croak "couldn't write to socket";

    my $line = $self->_check_line($socket->getline);
    $line =~ /^Client Error:/
      and croak "$line: $command";

    $line eq 'START'
      or return $line;

    my @lines;
    while (1) {
        $line = $self->_check_line($socket->getline);
        $line eq 'END'
          and last;
        push @lines, $line;
    }

    return @lines;
}

sub _check_line {
    my ($self, $line) = @_;
    if (!defined $line) {
        my $e = $!;
        if (any { $_ } ( $!{EWOULDBLOCK}, $!{EAGAIN}, $!{ETIMEDOUT} )) {
            $e = strerror(ETIMEDOUT);
            $self->disconnect;
        }
        undef $!;
        croak $e;
    }
    $line =~ s/$CR?$LF?$//;    
    return $line;
}

1;


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl client to the bloomd server",
  "AUTHOR" => "Damien \"dams\" Krotkine",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601",
    "Test::Exception" => 0,
    "Test::More" => 0,
    "Test::TCP" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Bloomd-Client",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Bloomd::Client",
  "PREREQ_PM" => {
    "Config" => 0,
    "IO::Socket::INET" => 0,
    "List::MoreUtils" => 0,
    "Method::Signatures" => 0,
    "Moo" => 0,
    "Type::Tiny" => "0.026"
  },
  "VERSION" => "0.25",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



